packages/coinstac-client-server/src/routes.js

```diff
@@ -0,0 +1,182 @@
+  }
+
+  const apiServer = `http://${process.env.API_SERVER_HOSTNAME}:${process.env.API_SERVER_PORT}`;
+
+  try {
+    await axios.post(`${apiServer}/authenticateByToken`, null, { headers: { Authorization: authorization } });
+    return true;
+  } catch {
+    return h.response('Invalid token.').code(400).takeover();
+  }
+};
+
+module.exports = (server) => {
+  return [
+    {
+      method: 'POST',
+      path: '/pullImages',
+      options: {
+        pre: [
+          { method: validateToken },
+        ],
+        handler: async (req, h) => {
+          try {
+            const { payload: { images } } = req;
+
+            await pullImagesFromList(images);
+            return h.response('Pulled images.').code(200);
+          } catch (error) {
+            return h.response(error.message).code(400);
+          }
+        },
+      },
+    },
+    {
+      method: 'POST',
+      path: '/removeImages',
+      options: {
+        pre: [
+          { method: validateToken },
+        ],
+        handler: async (req, h) => {
+          try {
+            const { payload: { images } } = req;
+
+            await removeImagesFromList(images);
+            return h.response('Removed images.').code(200);
+          } catch (error) {
+            return h.response(error.message).code(400);
+          }
+        },
+      },
+    },
+    {
+      method: 'POST',
+      path: '/startPipeline/{clientId}',
+      options: {
+        pre: [
+          { method: validateToken },
+        ],
+        handler: async (req, h) => {
+          try {
+            const { payload: { run }, params: { clientId } } = req;
+            const { runId } = run;
+
+            let subscribed = false;
+            const dataQueue = [];
+
+            const subscriptionURL = `/pipelineResult/${runId}`;
+
+            server.subscription(subscriptionURL, {
+              onSubscribe: () => {
+                subscribed = true;
+
+                dataQueue.forEach((data) => {
+                  server.publish(subscriptionURL, data);
+                });
+              },
+              onUnsubscribe: () => {
+                subscribed = false;
+              },
+            });
+
+            const remotePipelineManager = await getDockerManager(clientId);
+
+            const computationImageList = run.pipelineSnapshot.steps
+              .map(step => step.computations
+                .map(comp => comp.computation.dockerImage))
+              .reduce((acc, val) => acc.concat(val), []);
+
+            await pullImagesFromList(computationImageList);
+
+            const managerResponse = remotePipelineManager.startPipeline({
+              clients: run.clients,
+              spec: run.pipelineSnapshot,
+              runId,
+              timeout: run.timeout,
+            });
+
+            const { pipeline, result, stateEmitter } = managerResponse;
+
+            function publishData(data) { // eslint-disable-line no-inner-declarations
+              if (subscribed) {
```

Comment:
Are you sure this is working as expected? My intuition is that `subscribed` is being closured in `true` and the else is never being hit, and wouldn't change if the client unsubbed mid stream

Reply:
The purpose of this code is to cache the messages until client is subscribed.
once the client is connected, it will push all the messages in cache and then, it starts sending new messages without caching it.

I checked it on my side and worked, @rssk